lib/cn/cn_tree.c

```diff
@@ -1855,49 +1919,170 @@ cn_comp_cleanup(struct cn_compaction_work *w)
+    /* CNDB: Log kvset add records.
+     */
+    err = cndb_record_kvset_add(cndb, tx, w->cw_tree->cnid, km.km_nodeid,
+                                &km, ss->ss_kvsetid, km.km_hblk.bk_blkid,
+                                mblks->kblks.n_blks, (uint64_t *)mblks->kblks.blks,
+                                mblks->vblks.n_blks, (uint64_t *)mblks->vblks.blks, &cookie);
+    if (err)
         goto done;
 
-    /* Punt if job on head of list is not done or another thread is already committing it. */
-    if (!atomic_read(&w->cw_rspill_done) || atomic_read(&w->cw_rspill_commit_in_progress)) {
-        w = 0;
+    err = cn_mblocks_commit(w->cw_mp, 1, mblks, CN_MUT_OTHER);
+    if (err)
+        goto done;
+
+    err = kvset_open(w->cw_tree, ss->ss_kvsetid, &km, &kvset);
+    if (err)
         goto done;
-    }
 
-    /* Job on head of spill completion list is ready to be processed.
-     * - Set "commit_in_progress" status, but leave on list until commit is done.
-     * - If the node is wedged, it means an earlier job has failed, in
-     *   which case we force failure on this job to prevent out of
-     *   order completion.
-     * - If the node is not wedged, and this job has failed then it
-     *   will cause the node to be wedged, but this will be handled
-     *   later to catch downstream errors.
+    /* CNDB: Ack kvset add record.
      */
+    err = cndb_record_kvset_add_ack(cndb, tx, cookie);
+    if (err)
+        goto done;
 
-    atomic_set(&w->cw_rspill_commit_in_progress, 1);
+    cn_comp_kvset_append(w, ss->ss_node, kvset);
 
-    if (ev(tree->ct_rspills_wedged && !w->cw_err)) {
-        w->cw_err = merr(ESHUTDOWN);
-        w->cw_canceled = true;
-    }
+    blk_list_free(&ss->ss_mblks.kblks);
+    blk_list_free(&ss->ss_mblks.vblks);
 
 done:
-    mutex_unlock(&tree->ct_rspills_lock);
+    w->cw_t4_commit = get_time_ns();
+
+    if (err) {
+        cndb_record_nak(cndb, tx);
+        kvset_put_ref(kvset);
+    }
+
+    return err;
+}
+
+static merr_t
+cn_node_apply_subspill(struct subspill *ss)
+{
+    struct cn_compaction_work *w = ss->ss_work;
+    merr_t err;
+
+    err = cn_subspill_commit(ss);
+    if (err)
+        return err;
+
+    if (ss->ss_added) {
+        w->cw_output_nodev[0] = ss->ss_node;
+        w->cw_checkpoint(w);
+    }
 
-    return w;
+    return 0;
+}
+
+static merr_t
+cn_comp_spill(struct cn_compaction_work *w)
+{
+    struct subspill *ss;
+    struct spillctx *sctx;
+    struct cn_tree *tree = w->cw_tree;
+    struct route_node *rtn = 0;
+    uint cnum = 0;
+    merr_t err;
+
+    err = cn_spill_init(w, &sctx);
+    if (err)
+        return err;
+
+    while (!err) {
+        struct cn_tree_node *node;
+        void *lock;
+        struct kvset_list_entry *le;
+        uint64_t node_dgen;
+        unsigned char ekey[HSE_KVS_KEY_LEN_MAX];
+        uint eklen;
+
+        rmlock_rlock(&tree->ct_lock, &lock);
+
+        rtn = rtn ? route_node_next(rtn) : route_map_first_node(tree->ct_route_map);
```

Comment:
At one point we were going to get to the next node based on the last key from the previous subspill.  But I guess we can't do that now because we need to visit all nodes to bump the `sgen`.  Is that the only reason?

This is a bummer because it means spills that touch only one leaf still have to visit all leaves.  We need to noodle on this a bit to see if there's a better way.   Not to hold up this PR, but for a future improvement.

Does this mean the spills that don't touch a node, but have to increment the `sgen` have to be queued so the `sgen` is not bumped prematurely?

Reply:
Yeah, this took me a while to figure out when I was plumbing the locking to allow splits between rspill node visits.  I think we can improve this, but would rather do that after my split locking goes in...